### PR TITLE
Creates sdk-testing and adds ActorClientStub class.

### PR DESCRIPTION
**Pull request by artursouza**
# Description

Creates a new project: sdk-testing.
Adds ActorClientStub class for easy mocking of ActorClient class.

## Issue reference

We strive to have all PR being opened based on an issue, where the problem or feature have been discussed prior to implementation.

Please reference the issue this PR will close: #543 

Created new project (sdk-testing) to be used in #541 

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [X] Code compiles correctly
* [X] Created/updated tests
* [ ] Extended the documentation

### CODE REVIEW

**Comment by artursouza** [comment]
@mthmulders  @TomCools It would be great to have your feedback on this PR.

**Comment by mthmulders** [comment]
> @mthmulders @TomCools It would be great to have your feedback on this PR.

I will go through it next week.

**Comment by TomCools** [comment]
> > @mthmulders @TomCools It would be great to have your feedback on this PR.
> 
> I will go through it next week.

Idem Ditto 😄... weekend first!

File: sdk-testing/src/main/java/io/dapr/actors/client/ActorClientStub.java
Comment (TomCools) [inline sdk-testing/src/main/java/io/dapr/actors/client/ActorClientStub.java:12]: This is an incorrect assumption. Mockito allows you to mock Classes as well... as long they aren't _final_.
The reason for that behavior is because **Mockito will create a subclass at runtime** (see screenshot, you can see in the debugger that the instance "mock" is actually an automatically generated subclass of what you are mocking (ActorClient in this case).. So the users of the SDK don't actually need this stub implementation to help them test.

image

**Review by TomCools — COMMENTED** [review]
I think we more clearly need to define what kinds of features we would want in an "sdk-testing" module. Adding stubs like the one that was added now doesn't add much on top of Mockito. If we want to support other testing libraries... that's something else, but i'd argue that Mockito is pretty standard.

I do however see a lot of value in the following kind of test-utilities that might be placed in this module.

-  A Testing sidecar, and a way to spin up other "components + sidecar".
- These other components could be custom wrapper this module provides in the form of
   - For HTTP: "Sidecar + WireMock"
   - For Pubsub: "Sidecar + ?"

@mthmulders you have any more ideas on this?

File: sdk-testing/src/test/java/io/dapr/actors/client/ActorClientStubTest.java
Comment (mthmulders) [inline sdk-testing/src/test/java/io/dapr/actors/client/ActorClientStubTest.java:42]: Passing a diamond operator makes it clear for the compiler that you're using the generic type. You then leave it to the compiler to infer that generic type.

```suggestion
    ActorProxyBuilder<MyActor> builder = new ActorProxyBuilder<>(MyActor.class, mock);
```

File: sdk-testing/pom.xml
Comment (mthmulders) [inline sdk-testing/pom.xml:26]: JCenter has been [closed down](https://jfrog.com/blog/into-the-sunset-bintray-jcenter-gocenter-and-chartcenter/) I believe. I would not expect new references to be introduced (and I would even expect old references to disappear someday - but that should be a separate ticket, probably?)

**Review by mthmulders — COMMENTED** [review]
I am not yet familiar with Actors in Dapr, so I'm having a hard time understanding what the added value of these changes is.

In general, I agree with @TomCools, it would probably be a lot more useful if we could offer "ready to roll" sidecars for testing. I'm thinking something like [Test Containers](https://www.testcontainers.org/). We might even use them to test the Spring Boot SDK.

File: sdk-testing/src/main/java/io/dapr/actors/client/ActorClientStub.java
Comment (artursouza) [reply sdk-testing/src/main/java/io/dapr/actors/client/ActorClientStub.java:12]: Awesome! I will update the comment. On the other hand, we still need to make the method public here. I was keeping it package protected to make sure the App does not invoke any method in ActorClient and only used internally.

File: sdk-testing/pom.xml
Comment (artursouza) [reply sdk-testing/pom.xml:26]: Yes, that needs to be addressed. Too bad. We had to migrate to jcenter due to another mirror being closed down before. It seems to be a thing now for public Maven repositories.

Comment (artursouza) [reply sdk-testing/pom.xml:26]: Tracking here: https://github.com/dapr/java-sdk/issues/608

**Comment by artursouza** [comment]
> I think we more clearly need to define what kinds of features we would want in an "sdk-testing" module. Adding stubs like the one that was added now doesn't add much on top of Mockito. If we want to support other testing libraries... that's something else, but i'd argue that Mockito is pretty standard.
> 
> I do however see a lot of value in the following kind of test-utilities that might be placed in this module.
> 
> * A Testing sidecar, and a way to spin up other "components + sidecar".
> * These other components could be custom wrapper this module provides in the form of
>   
>   * For HTTP: "Sidecar + WireMock"
>   * For Pubsub: "Sidecar + ?"
> 
> @mthmulders you have any more ideas on this?



We have an issue to create this "testing SDK" project: https://github.com/dapr/java-sdk/issues/541

yes, the idea is that it can help apps to write integration tests with actual Dapr sidecar running. We have that code already in our internal testing package (sdk-tests), so the idea is to create an user friendly API.

**Comment by mthmulders** [comment]
A second line of thought on this: I typically try to avoid mocking "things I don't own". If I were building applications with Dapr, that would mean I would avoid mocking Dapr internals. I'd rather not have my test depend on the internals of an SDK I use.

I'd prefer a test that proofs the behaviour of my production code over a test that proofs I invoke the right methods of a third-party class. The latter may change as that SDK evolves. The former should hardly ever change.

By providing convenient mocks of our own internals, I feel we're encouraging consumers to do the former.

**Comment by artursouza** [comment]
> I like the idea of this as well, but it'd be good to see it in action. We have the actor example in the Java SDK already. Do you think we could see how this would add a unit test/integration test to that as an example? I think that would help show what value this adds, or, show us that we're OK with just Mockito for now.


The unit test in this code shows that.

**Comment by artursouza** [comment]
> A second line of thought on this: I typically try to avoid mocking "things I don't own". If I were building applications with Dapr, that would mean I would avoid mocking Dapr internals. I'd rather not have my test depend on the internals of an SDK I use.
> 
> I'd prefer a test that proofs the behaviour of my production code over a test that proofs I invoke the right methods of a third-party class. The latter may change as that SDK evolves. The former should hardly ever change.
> 
> By providing convenient mocks of our own internals, I feel we're encouraging consumers to do the former.


I've seen this approach before and it probably means creating Facade interfaces to abstract 3rd party interfaces. This is not  necessarily a universal practice. We already encourage mocking of Dapr client by providing it as an interface. If a developer wants to avoid that, he then should come up with his/her own interfaces to avoid exposing any direct interaction with Dapr.

To a greater extend, Dapr is that abstraction already - avoids applications to implement against specific interfaces of their cloud components by providing a common interface.

**Comment by artursouza** [comment]
/ping @halspang @TomCools @mthmulders 

Code freeze is today, so I assume we will not include this in this release unless we get some consensus on this. As a maintainer, I seek consensus as much as possible - so will give everyone some time to agree on a path forward here.

**Review by halspang — APPROVED** [review]
Looks good to me!

**Comment by artursouza** [comment]
I talked to @rynowak about this. I will try a solution where the user provides a mock implementation of their actor interface and we offer an ActorClient implementation that calls into the in-memory implementation. So users just mock their actor interface and not the method from ActorClient. I will try that in the next release, since code freeze is today. Thanks for playing.

**Review by mukundansundar — APPROVED** [review]
lgtm

**Pull request closed without merging**
